### PR TITLE
Bring closer to pairity with supabase/gotrue-js

### PR DESCRIPTION
**Pull request by fedden**
Adds a bunch of tests
More closely matches the structure of the js library, which will help us keep on top of future changes.

### CODE REVIEW

**Review by fedden — COMMENTED** [review]
@J0 let me know what you think!

File: gotrue/client.py [hunk 1]
```diff
@@ -3,75 +3,238 @@
...
+        self.auto_refresh_token = auto_refresh_token
+        self.persist_session = persist_session
+        self.local_storage: Dict[str, Any] = {}
+        self.api = GoTrueApi(url=url, headers=headers, cookie_options=cookie_options)
+        self._recover_session()
+
+    def sign_up(self, email: str, password: str):
+        """Creates a new user.
+
+        Paramters
```
Comment (J0) [inline gotrue/client.py:66]: /nit parameters
```suggestion
        Parameters
```

**Review by J0 — APPROVED** [review]
LGTM. Thanks for doing this!

File: gotrue/api.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        """Initialise API class."""
+        self.url = url
+        self.headers = headers
+        self.cookie_options = {**COOKIE_OPTIONS, **cookie_options}
+
+    def sign_up_with_email(self, email: str, password: str) -> Dict[str, Any]:
+        """Creates a new user using their email address
+
+        Parameters
+        ---------
```
[the remark this answers is not in the document]
Comment (fedden) [reply gotrue/api.py:27]: Yes generally we do. I internally have the rule that the worst thing I can do is write a single-line docstring describing what the code does. The best thing I can do is describing all parameters, returns, example usage, etc etc - I rarely do this though.

[the remark this answers is not in the document]
Comment (fedden) [reply gotrue/api.py:27]: I'll add a return statement but guess I'm saying there always is more that can be done!

File: gotrue/api.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        """
+        request = requests.post(
+            f"{self.url}/token?grant_type=refresh_token",
+            json.dumps({"refresh_token": refresh_token}),
+            headers=self.headers,
+        )
+        return to_dict(request)
+
+    def set_auth_cookie(req, res):
+        """Stub for pairty with JS api."""
```
[the remark this answers is not in the document]
Comment (fedden) [reply gotrue/api.py:166]: Thanks!

File: gotrue/client.py (hunk 1 above)
Comment (fedden) [reply gotrue/client.py:66]: Thanks!

File: gotrue/api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fedden) [reply gotrue/api.py]: But thanks for the note, great point, and I've added return statements to all public api.py methods

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -9,15 +9,16 @@ on:
...
     name: Test / OS ${{ matrix.os }} / Node ${{ matrix.node }}
     strategy:
       matrix:
         os: [ubuntu-latest]
         python-version: [3.6, 3.7, 3.8, 3.9]
-
-
     runs-on: ${{ matrix.os }}
-
+    # TODO(fedden): We need to discuss these steps: We could just use a test-supabase instance or we could update the docker image and use that for the tests.
```
[the remark this answers is not in the document]
Comment (fedden) [reply .github/workflows/ci.yml:21]: Yea agreed, this is what I've been doing so far in my supabase-py branch and this fork of gotrue-py!

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -9,15 +9,16 @@ on:
 
 jobs:
   test:
+    env:
+      SUPABASE_TEST_URL: "https://tfsatoopsijgjhrqplra.supabase.co"
```
[the remark this answers is not in the document]
Comment (fedden) [reply .github/workflows/ci.yml:13]: Sounds good to me thanks for the suggestion

[the remark this answers is not in the document]
Comment (fedden) [reply .github/workflows/ci.yml:13]: I think this CI yaml needs an overhaul anyway! Lets take care of it in a future PR

[the remark this answers is not in the document]
Comment (fedden) [reply .github/workflows/ci.yml:13]: I've removed the envs for now

File: gotrue/client.py [hunk 6]
```diff
@@ -3,75 +3,238 @@
...
 import json
-from urllib.parse import quote
+import re
+import uuid
+from typing import Any, Callable, Dict, Optional
 
-HTTPRegexp = "/^http://"
-defaultApiURL = "/.netlify/identity"
+from gotrue.api import GoTrueApi
+from gotrue.lib.constants import GOTRUE_URL, STORAGE_KEY
```
[the remark this answers is not in the document]
Comment (fedden) [reply gotrue/client.py:14]: Just copying the JS clients structure to ensure it's easy to compare the two clients, but thanks!

**Comment by fedden** [comment]
Okay @J0 I've addressed all PR comments - thanks for leaving the review. Merge when you are ready

File: gotrue/api.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (J0) [reply gotrue/api.py]: Yeah, your internal rule makes sense -- don't think we should need to re-describe the code.  Thanks for adding the  single line return statements.

File: .github/workflows/ci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (J0) [reply .github/workflows/ci.yml]: cool, thanks!

**Pull request merged**